### PR TITLE
`isset()` narrows string-key in int-keyed-array to numeric-string

### PR DESCRIPTION
**Pull request by staabm**
initial steps to make `isset()` implementation even more similar to `array_key_exists` so we can in the end hopefully share the implementation

commit 1: plain 1:1 refactoring, no implementation changes
commit 2: the fix as the PR topic suggests

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -820,47 +821,8 @@ public function specifyTypesInCondition(
...
-									false,
-									new UnionType([
-										new ArrayType(new MixedType(), new MixedType()),
-										new ObjectWithoutClassType(),
-										new ResourceType(),
-									]),
-								);
-							}
-
+						$narrowedKey = AllowedArrayKeysTypes::narrowOffsetKeyType($varType, $dimType);
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:824]: as a first step I think its easier to have a static method for this.

after we get more similar with the `array_key_exists` impl we will see whether it makes sense to move this method into the `ConstantScalarType`-interface, as we discussed in dresden

File: tests/PHPStan/Analyser/nsrt/bug-11716.php [hunk 2]
```diff
@@ -59,21 +60,42 @@ function narrowKey($mixed, string $s, int $i, array $generalArr, array $arr): vo
...
+	if (isset($intKeyedArr[$i])) {
 		assertType('int', $i);
 	} else {
 		assertType('int', $i);
 	}
 	assertType('int', $i);
 
-	if (isset($arr[$s])) {
+	if (isset($intKeyedArr[$s])) {
+		assertType('numeric-string', $s);
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/bug-11716.php]: this assertion is the interessting one, which got more precise with this PR

**Comment by staabm** [comment]
~~I am not sure whether this CI errors are related to my PR, or whether these are triggered by recent BC breaks on the 2.x branch.~~ 

-> it seems to be caused by some github infrastructure thing -> works again

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**